engine/src/main/java/io/zeebe/engine/processing/bpmn/gateway/ParallelGatewayProcessor.java

```diff
@@ -31,51 +29,29 @@ public ParallelGatewayProcessor(final BpmnBehaviors behaviors) {
   }
 
   @Override
-  public void onActivating(final ExecutableFlowNode element, final BpmnElementContext context) {
+  public void onActivate(final ExecutableFlowNode element, final BpmnElementContext context) {
     // the joining of the incoming sequence flows into the parallel gateway happens in the
-    // sequence flow processor
-
-    // the activating event of the parallel gateway is written when all incoming sequence flows are
-    // taken
-
-    stateTransitionBehavior.transitionToActivated(context);
-  }
-
-  @Override
-  public void onActivated(final ExecutableFlowNode element, final BpmnElementContext context) {
-
-    stateTransitionBehavior.transitionToCompleting(context);
-  }
-
-  @Override
-  public void onCompleting(final ExecutableFlowNode element, final BpmnElementContext context) {
-
-    stateTransitionBehavior.transitionToCompleted(context);
-  }
-
-  @Override
-  public void onCompleted(final ExecutableFlowNode element, final BpmnElementContext context) {
+    // sequence flow processor. The activating event of the parallel gateway is written when all
+    // incoming sequence flows are taken
+    final var activated = stateTransitionBehavior.transitionToActivated(context);
+    final var completing = stateTransitionBehavior.transitionToCompleting(activated);
+    final var completed = stateTransitionBehavior.transitionToCompleted(completing);
     // fork the workflow processing by taking all outgoing sequence flows of the parallel gateway
-
-    stateTransitionBehavior.takeOutgoingSequenceFlows(element, context);
-
-    stateBehavior.consumeToken(context);
-    stateBehavior.removeElementInstance(context);
+    stateTransitionBehavior.takeOutgoingSequenceFlows(element, completed);
   }
 
   @Override
-  public void onTerminating(final ExecutableFlowNode element, final BpmnElementContext context) {
-
-    stateTransitionBehavior.transitionToTerminated(context);
+  public void onComplete(final ExecutableFlowNode element, final BpmnElementContext context) {
+    throw new UnsupportedOperationException(
+        String.format(
+            "Expected to explicitly process complete, but gateway %s has no wait state",
```

Comment:
I think this error message is not clear enough. Neither for a user nor for a developer which doesn't have the full background of the changes we are currently making. Please see if you can improve this. Either the error message or adding comments for developers.

Reply:
@pihme Can you explain what is not clear about it? Then I know better what is missing.

Reply:
Well, it is not clear what has happened. Why this is unexpected. And what to do about it.

I think what you are trying to say is something like this:

Received `onComplete` command. This is unexpected. It is unexpected because this is a parallel gateway which has no explicit on complete step. Instead, the activities that normally fall into the ´onComplete´ stage are rolled into the handling of the ´activateElement`command. 

If a user. There is nothing you can do, it's probably a bug. Please open a ticket.

If a developer, find out who sent the on complete command. Don't send it anymore.

Reply:
I'm not sure I agree with putting so much info in the message of an exception. Most of this info is actually already implicitly available. For example, exceptions are always bugs in our engine and will blacklist the instance, if there was something the user could do we should've raised an incident instead.

Perhaps the word wait state is not clear enough. What do you think of something like this:
```suggestion
            "Expected to explicitly process complete, but gateway %s has already been completed on processing activate",
```

Reply:
Sure, works for me.